File: Stack and Queues/649. Dota2 Senate.cpp
class Solution {
public:
    string predictPartyVictory(string senate) {
        int net = 0;
        vector<char> stack(senate.begin(), senate.end());


        while (true) {
            vector<char> new_stack;
            bool should = false;
            for (char s: stack) {
                if (s == 'D') {
                    net++;
                    if (net > 0) {
                        new_stack.push_back(s);
                    } else {
                        should = true;
                    }
                } else {
                    net--;
                    if (net < 0) {
                        new_stack.push_back(s);
                    } else {
                        should = true;
                    }
                }
            }

            stack = new_stack;
            if (!should) {
                break;
            }

        }

        if (net > 0) return "Dire";

        return "Radiant";

    }
};

File: Stack and Queues/649. Dota2 Senate.py
class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        # play the strategy for your own party
        # ending score should be equals to the length of the string
        net = 0
        stack = list(senate)

        while True:
            new = []
            should = False
            for s in stack:
                if s == 'D':
                    net += 1
                    if net > 0:
                        new.append(s)
                    else:
                        should = True

                elif s == 'R':
                    net -= 1
                    if net < 0:
                        new.append(s)
                    else:  # that means I deleted an element
                        should = True

            stack = new
            if not should:
                break

        if net > 0:
            return "Dire"

        return "Radiant"

